src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/PairedEnds.java

```diff
@@ -1,130 +1,441 @@
 package org.broadinstitute.hellbender.utils.read.markduplicates;
 
+import com.esotericsoftware.kryo.DefaultSerializer;
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
 import htsjdk.samtools.SAMFileHeader;
+import org.broadinstitute.hellbender.utils.Utils;
 import org.broadinstitute.hellbender.utils.read.GATKRead;
 import org.broadinstitute.hellbender.utils.read.ReadUtils;
 
 /**
  * Struct-like class to store information about the paired reads for mark duplicates.
  */
-public class PairedEnds implements OpticalDuplicateFinder.PhysicalLocation {
-  private GATKRead first, second;
-
-  // Information used to detect optical dupes
-  public short readGroup = -1;
-  public short tile = -1;
-  public short x = -1, y = -1;
-  public short libraryId = -1;
-
-  PairedEnds(final GATKRead first) {
-    this.first = first;
-  }
-
-  public static PairedEnds of(final GATKRead first) {
-    return new PairedEnds(first);
-  }
-
-  public PairedEnds and(final GATKRead second) {
-    if (second != null &&
-        ReadUtils.getStrandedUnclippedStart(first) > ReadUtils.getStrandedUnclippedStart(second)) {
-      this.second = this.first;
-      this.first = second;
-    } else {
-      this.second = second;
+public abstract class PairedEnds extends MarkDuplicatesSparkData {
```

Comment:
It's confusing that we call this thing PairedEnds when only 1 implementation is a paired end.  I'm not sure what we can call it instead though....

Reply:
I would convert this to an interface since it has no common fields or implementation

Reply:
I think we should consider creating a markduplicatesspark package, and moving these and the record classes as top level classes there

Reply:
Hmn. maybe not convert it to an interface since I'm not sure if it make sense to have ones that aren't MakrDuplicateSparkRecords.